### PR TITLE
Upgrade geth to v1.14

### PR DESCRIPTION
**Pull request by wsodsong**
## Description

This PR upgrades geth to version 1.14. The change includes
- support new state db facades
- balance type change to *uint256
- change message type to core.Message

Changes which will be resolved in a later PR
- #1121 Re-enable equality check
- #1122 Support new statedb facades in tracer, profiler and stochastic tool
- #1103 Carmen transient storage

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactoring (changes that do NOT affect functionality)
- [x] Adds or updates testing

### CODE REVIEW

File: rpc/evm.go [hunk 1]
```diff
@@ -112,7 +111,7 @@ func newTxArgs(params map[string]interface{}) ethapi.TransactionArgs {
 }
 
 // newEVM creates new instance of EVM with given parameters
-func (e *EvmExecutor) newEVM(msg eth.Message, hashErr *error) *vm.EVM {
+func (e *EvmExecutor) newEVM(msg core.Message, hashErr *error) *vm.EVM {
```
Comment (petr-hanzl) [inline rpc/evm.go:114]: Pass a pointer to `core.Message` here, you have to create it anyway on line `144`

File: state/carmen.go [hunk 2]
```diff
@@ -104,6 +106,11 @@ func (s *carmenStateDB) CreateAccount(addr common.Address) {
 	s.txCtx.CreateAccount(carmen.Address(addr))
 }
 
+func (s *carmenStateDB) CreateContract(addr common.Address) {
+	panic("CreateContract not implemented")
+	return
+}
```
Comment (petr-hanzl) [inline state/carmen.go:112]: Does this need to be implemented soon? Should we create an issue?

File: state/carmen.go [hunk 3]
```diff
@@ -160,6 +164,21 @@ func (s *carmenStateDB) SetState(addr common.Address, key common.Hash, value com
...
+
+func (s *carmenStateDB) GetTransientState(addr common.Address, key common.Hash) common.Hash {
+	panic("GetTransientState not implemented")
+	return common.Hash{}
+}
+
+func (s *carmenStateDB) SetTransientState(addr common.Address, key common.Hash, value common.Hash) {
+	panic("SetTransientState not implemented")
+	return
+}
```
Comment (petr-hanzl) [inline state/carmen.go:180]: This is already implemented [here](https://github.com/Fantom-foundation/Carmen/blob/ad1657431037430a60f7456088947e5bfd16972a/go/carmen/transaction.go#L115)

File: state/carmen.go [hunk 4]
```diff
@@ -160,6 +164,21 @@ func (s *carmenStateDB) SetState(addr common.Address, key common.Hash, value com
 	s.txCtx.SetState(carmen.Address(addr), carmen.Key(key), carmen.Value(value))
 }
 
+func (s *carmenStateDB) GetStorageRoot(addr common.Address) common.Hash {
+	panic("GetStorageRoot not implemented")
+	return common.Hash{}
+}
```
Comment (petr-hanzl) [inline state/carmen.go:170]: fyi: We agreed that we actually dont need this func. We only need an information whether the storage is empty or not. The func will be `HasEmptyStorage(addr) bool`. A PR with this change is currently under review.

File: state/carmen_test.go [hunk 5]
```diff
@@ -775,7 +773,7 @@ func TestCarmenState_BulkloadOperations(t *testing.T) {
 				switch {
 				case operationType == 1:
 					// set balance
-					newBalance := big.NewInt(int64(GetRandom(0, 1000*5000)))
+					newBalance := uint256.NewInt(uint64(GetRandom(0, 1000*5000)))
```
Comment (petr-hanzl) [inline state/carmen_test.go:776]: nit: Maybe make `GetRandom` return a uint64?

File: state/memory.go [hunk 6]
```diff
@@ -217,12 +223,28 @@ func (db *inMemoryStateDB) SetState(addr common.Address, key common.Hash, value
...
+	panic("GetStorageRoot not implemented")
+}
+
+func (db *inMemoryStateDB) GetTransientState(addr common.Address, key common.Hash) common.Hash {
+	panic("GetTransientState not implemented")
+}
+
+func (db *inMemoryStateDB) SetTransientState(addr common.Address, key common.Hash, value common.Hash) {
+	panic("SetTransientState not implemented")
+}
```
Comment (petr-hanzl) [inline state/memory.go:236]: I think its time to merge [this](https://github.com/Fantom-foundation/Aida/pull/1103) PR so you can use the implementation here. I wouldnt like to break your branch so you can just merge that branch into this one.

**Review by cabrador — CHANGES_REQUESTED** [review]

File: utils/default_config.go [hunk 7]
```diff
@@ -97,7 +97,8 @@ func createConfigFromFlags(ctx *cli.Context) *Config {
 		SnapshotDepth:          getFlagValue(ctx, SnapshotDepthFlag).(int),
 		StateDbSrc:             getFlagValue(ctx, StateDbSrcFlag).(string),
 		StateDbSrcDirectAccess: getFlagValue(ctx, SrcDbOverwriteFlag).(bool),
-		StateValidationMode:    EqualityCheck,
+		// TODO re-enable equality check once supported in Carmen
```
Comment (petr-hanzl) [inline utils/default_config.go:100]: nit: maybe create an issue for this todo?

File: tracer/operation/operation_test.go [hunk 8]
```diff
@@ -273,6 +274,26 @@ func (s *MockStateDB) GetSubstatePostAlloc() txcontext.WorldState {
...
+	panic("GetStorageRoot not supported in mock")
+}
+
+func (s *MockStateDB) SetTransientState(addr common.Address, key common.Hash, value common.Hash) {
+	panic("SetTransientState not supported in mock")
+}
+
+func (s *MockStateDB) GetTransientState(addr common.Address, key common.Hash) common.Hash {
+	panic("GetTransientState not supported in mock")
+}
```
Comment (petr-hanzl) [inline tracer/operation/operation_test.go:295]: You can use [this](https://github.com/Fantom-foundation/Aida/pull/1103) PR as inspiration

File: tracer/operation/hassuicided.go [hunk 9]
```diff
@@ -59,7 +59,8 @@ func (op *HasSuicided) Write(f io.Writer) error {
 func (op *HasSuicided) Execute(db state.StateDB, ctx *context.Replay) time.Duration {
 	contract := ctx.DecodeContract(op.Contract)
 	start := time.Now()
-	db.HasSuicided(contract)
+	//TODO change function name
```
Comment (petr-hanzl) [inline tracer/operation/hassuicided.go:62]: nit: todo?

**Review by cabrador — COMMENTED** [review]
I think we should merge the [PR](https://github.com/Fantom-foundation/Aida/pull/1103) with TransientStorage and then use it here.

File: tracer/operation/operation_test.go (hunk 8 above)
Comment (wsodsong) [reply tracer/operation/operation_test.go:295]: I plan to add new operations for tracer and profiler in a separate PR. Some of them has already been done by your PR.

File: state/carmen.go (hunk 3 above)
Comment (wsodsong) [reply state/carmen.go:180]: Yes, however, I have to refer to the older version of Carmen which matches the one in Sonic to resolve dependency issue. The new facades of carmen are not available.

File: state/carmen.go (hunk 4 above)
Comment (wsodsong) [reply state/carmen.go:170]: GetStorageRoot returns a hash. What should be returned if `HasEmptyStorage(addr) bool` is used instead?

File: state/memory.go (hunk 6 above)
Comment (wsodsong) [reply state/memory.go:236]: No right now. As mentioned in another comment, I have to use an older version of carmen without `TransientState`. This PR https://github.com/Fantom-foundation/Sonic/pull/144 has to be merged first.

File: state/carmen.go (hunk 4 above)
Comment (petr-hanzl) [reply state/carmen.go:170]: I think the correct way to return an empty hash for `true` and non-empty for `false`. Afaik @HerbertJordan said that the way this func is utilized is only check for emptiness.

File: state/memory.go (hunk 6 above)
Comment (petr-hanzl) [reply state/memory.go:236]: ic

File: state/carmen.go (hunk 4 above)
Comment (HerbertJordan) [reply state/carmen.go:170]: That's right. `GetStorageRoot` is needed [here](https://github.com/Fantom-foundation/go-ethereum-sonic/blob/2400937cc3b1c370ea8735d7af86eac621e4eb70/core/vm/evm.go#L480) only to be compared with either the zero-hash or the hash of an empty trie.

So, to implement this function it should be sufficient to call `HasEmptySotrage(addr)` and return `Hash{0}` if the result is `true` and `Hash{1}` if the result is `false`.

File: tracer/operation/hassuicided.go (hunk 9 above)
Comment (wsodsong) [reply tracer/operation/hassuicided.go:62]: An upgrade to tracer and profiler will be added in the next PR.

File: executor/transaction_processor.go [hunk 10]
```diff
@@ -184,16 +184,16 @@ func (s *TxProcessor) processRegularTx(db state.VmStateDB, block int, tx int, st
 
 	// if no prior error, create result and pass it to the data.
 	blockHash := common.HexToHash(fmt.Sprintf("0x%016d", block))
-	res = newTransactionResult(db.GetLogs(txHash, blockHash), msg, msgResult, err, evm.TxContext.Origin)
+	res = newTransactionResult(db.GetLogs(txHash, uint64(block), blockHash), msg, msgResult, err, evm.TxContext.Origin)
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply executor/transaction_processor.go:187]: Do yo mean `uint64`?

File: rpc/evm.go [hunk 11]
```diff
@@ -278,7 +277,7 @@ func (e *EvmExecutor) findHiLoCap() (uint64, uint64, uint64, error) {
 	// Recap the highest gas limit with account's available balance.
 	if feeCap.BitLen() != 0 {
 		balance := e.archive.GetBalance(*e.args.From) // from can't be nil
-		available := new(big.Int).Set(balance)
+		available := new(big.Int).Set(balance.ToBig())
```
Comment (HerbertJordan) [inline rpc/evm.go:280]:
```suggestion
		available := balance.ToBig()
```

File: state/carmen.go (hunk 2 above)
Comment (HerbertJordan) [reply state/carmen.go:112]: Please see #1119

File: state/carmen.go (hunk 4 above)
Comment (HerbertJordan) [reply state/carmen.go:170]: Please see #1119

File: state/carmen_test_enviroment.go [hunk 12]
```diff
@@ -90,13 +90,13 @@ func MakeRandomByteSlice(t *testing.T, bufferLength int) []byte {
 	return buffer
 }
 
-func GetRandom(rangeLower int, rangeUpper int) int {
+func GetRandom(rangeLower int, rangeUpper int) uint64 {
```
Comment (HerbertJordan) [inline state/carmen_test_enviroment.go:93]: Should the lower and upper limit of the range also be updated to `uint64` to match the result?

Odd results may be produced from calls like `GetRandom(-10,10)`.

**Review by HerbertJordan — APPROVED** [review]
Looks good to me, thanks for putting in all that effort to get this upgrade done!

File: cmd/aida-vm-adb/run_vm_adb_test.go [hunk 13]
```diff
@@ -420,16 +421,16 @@ func TestVmAdb_ValidationDoesNotFailOnValidTransaction_Sequential(t *testing.T)
...
-		archive.EXPECT().Prepare(gomock.Any(), 1),
+		archive.EXPECT().SetTxContext(gomock.Any(), 1),
 		archive.EXPECT().Snapshot().Return(15),
-		archive.EXPECT().GetBalance(gomock.Any()).Return(big.NewInt(1000)),
-		archive.EXPECT().SubBalance(gomock.Any(), gomock.Any()),
+		archive.EXPECT().GetBalance(gomock.Any()).Return(uint256.NewInt(1000)),
+		archive.EXPECT().SubBalance(gomock.Any(), gomock.Any(), gomock.Any()),
 		archive.EXPECT().RevertToSnapshot(15),
-		archive.EXPECT().GetLogs(common.HexToHash(fmt.Sprintf("0x%016d%016d", 2, 1)), common.HexToHash(fmt.Sprintf("0x%016d", 2))),
+		archive.EXPECT().GetLogs(common.HexToHash(fmt.Sprintf("0x%016d%016d", 2, 1)), uint64(2), common.HexToHash(fmt.Sprintf("0x%016d", 2))),
```
[the remark this answers is not in the document]
Comment (wsodsong) [reply cmd/aida-vm-adb/run_vm_adb_test.go:433]: I don't have a good answer for this. At Line 416, block 2 tx 1 is consumed and there is no information about Block 1 and no substate of block 1.  It looks to me that block 2 is being processed.

[the remark this answers is not in the document]
Comment (wsodsong) [reply cmd/aida-vm-adb/run_vm_adb_test.go:433]: This test runs archive at block 2 instead.  GetArchive state returns the state at the end of block 1 / beginning of block 2. GetLogs works as expected.

File: utils/default_config.go (hunk 7 above)
Comment (wsodsong) [reply utils/default_config.go:100]: see #1121

**Comment by wsodsong** [comment]
Waiting for results of manual tests.

File: state/carmen_test_enviroment.go (hunk 12 above)
Comment (wsodsong) [reply state/carmen_test_enviroment.go]: With all possible type conversions, I decided to avoid them and throw errors if invalid numbers are passed.

**Comment by wsodsong** [comment]
> Waiting for results of manual tests.

- [x] Aida-vm mainnet full range: passed
- [x] Aida-vm testnet full range: passed
- [x] Aida-vm-sdb testnet full range: passed

**Review by HerbertJordan — APPROVED** [review]
Thanks for all that work!

File: state/carmen.go [hunk 14]
```diff
@@ -160,6 +166,53 @@ func (s *carmenStateDB) SetState(addr common.Address, key common.Hash, value com
...
+	// Since Fantom has always had EIP-158, such accounts should not exist on the
+	// Fantom networks. Nevertheless, Carmen is implementing a check for the emptiness
+	// of the storage to be Ethereum compatible. Once this is offered, this function
+	// should be updated.
+
+	// Until Carmen offers a way to determine whether
+	// the storage of an account is empty or not, we return a zero hash here
+	// indicating that the storage is empty -- which corresponds to pre EIP-7610
+	// behavior.
+	// TODO: use Carmen's HasEmptyStorage function once available.
```
Comment (petr-hanzl) [inline state/carmen.go:202]: You can remove this todo and simply return empty hash. This [comment](https://github.com/Fantom-foundation/Carmen/pull/924#issuecomment-2180337715) explains why this scenario wont happen on our chain.

**Review by cabrador — APPROVED** [review]
Thanks for this!

**Pull request merged**